design/open/delta-diff.md

```diff
@@ -0,0 +1,186 @@
+
+We can trigger a diff binary from lakeFS using the `os.exec` package and get the output generated by that binary.
+This is almost where we want to be, except that we'll need to somehow validate that the lakeFS server and the executable are using the same "API version" to communicate, so that the output of the binary would match the expected one from lakeFS. In addition, I would like to decrease the amount of deserialization needed to be implemented to interpret the returned DTO (maintainability and extensibility-wise).
+
+### Chosen architecture
+
+#### The Microkernel/Plugin Architecture
+
+The Microkernel/Plugin architecture is composed of two entities: a single "core system" and multiple "plugins".  
+In our case, the lakeFS server act as the core system, and the different diff implementations, including the Delta diff implementation, will act as plugins.  
+We'll use `gRPC` as the transport protocol, which makes the language of choice almost immaterial (due to the use of protobufs as the transferred data format)
+as long as it's self-contained (theoretically it can also be not system-runtime-dependent but then the cost will be an added requirement for running lakeFS- runtime support).
+
+#### Plugin system consideration
+
+* The plugin system should be flexible enough such that it won't impose a language restriction, so that in the future we could have lakeFS users take part in plugins creation (not restricted to diffs).
+* The plugin system should support multiple OSs.
+* The plugin system should be easy enough to write plugins.
+
+#### Hashicorp Plugin system
+
+Hashicorp's battle-tested [`go-plugin`](https://github.com/hashicorp/go-plugin) system is a plugin system over RPC (used in Terraform, Vault, Consul, and Packer).
+Currently, it's only designed to work over a local network.  
+Plugins can be written and consumed in almost every major language. This is achieved by supporting [gRPC](http://www.grpc.io/) as the communication protocol.
+* The plugin system works by launching subprocesses and communicating over RPC (both `net/rpc` and `gRPC` are supported).
+  A single connection is made between any plugin and the core process. For gRPC-based plugins, the HTTP2 protocol handles [connection multiplexing](https://freecontent.manning.com/animation-http-1-1-vs-http-2-vs-http-2-with-push/).
+* The plugin and core system are separate processes, which means that a crash of a plugin won't cause the core system to crash (lakeFS in our case).
+
+![Microkernel architecture overview](diagrams/microkernel-overview.png)
+[(excalidraw file)](diagrams/microkernel-overview.excalidraw)
+
+---
+
+## Implementation
+
+![Delta Diff flow](diagrams/delta-diff-flow.png)
+[(excalidraw file)](diagrams/delta-diff-flow.excalidraw)
+
+#### DiffService
+
+The DiffService will be an internal component in lakeFS which will serve as the Core system.  
+In order to realize which diff plugins are available, we shall use new configuration values:
```

Comment:
- There is no manifest / configuration to map which plugins to load? from reading this section we have location and the specific delta diff information. Suggest you define the plugin configuration without sharing the location.
  Example:
  ```yaml
  plugins:
    diff:
      delta: <location to the delta diff plugin - full path to the binary needed to execute>    
  ```
- In case we like to map multiple plugins to the same format we should consider separating the plugins configuration and the use cases. Like:
  ```yaml
  diff:
    otf:
      delta:
        plugin: diff_delta_v2
  plugins:
    diff_delta_v1: <location to the delta diff plugin - full path to the binary needed to execute>    
    diff_delta_v2: <location to the delta diff plugin - full path to the binary needed to execute>    
  ```
- Another consideration - if we plan to use the plugin later also to support advance merge - we can just drop the `diff` from all the above examples.
- In case we would like to leverage more plugins for different use-cases (ex: merge) will it possible to share the same location for different plugin interfaces or we need to split / catalog them now?

Reply:
```yaml
diff:
  otf:
    delta:
      plugin: diff_delta_v2
plugins:
  diff_delta_v1: <location to the delta diff plugin - full path to the binary needed to execute>    
  diff_delta_v2: <location to the delta diff plugin - full path to the binary needed to execute>    
```
Really liked this one, although I would still want to have default values so that if the user didn't specify a plugin location in the manifest, lakeFS will try to figure it out using the latest known version of the plugin and try to locate it in some default path. For example, if the user didn't specify the location for `diff.delta.plugin`, lakeFS will try to locate the binary under `~/.lakefs/plugins`.

> * Another consideration - if we plan to use the plugin later also to support advance merge - we can just drop the diff from all the above examples.
> * In case we would like to leverage more plugins for different use-cases (ex: merge) will it possible to share the same location for different plugin interfaces or we need to split / catalog them now?

I think that it's a very possible scenario, yet I think that it might be preferable that each plugin would know how to do one thing. Anyway, we can decide upon it when we get there...